src/ucp/rndv/rndv_put.c

```diff
@@ -563,8 +569,20 @@ ucp_proto_rndv_put_mtype_probe(const ucp_proto_init_params_t *init_params)
     ucp_md_map_t mdesc_md_map;
     ucs_status_t status;
     size_t frag_size;
+    ucs_memory_type_t frag_mem_type;
 
-    status = ucp_proto_rndv_mtype_init(init_params, &mdesc_md_map, &frag_size);
+    if (init_params->rkey_config_key == NULL) {
+        frag_mem_type = UCS_MEMORY_TYPE_HOST;
+    } else {
+        /* Can initialize only the same fragment type as received in RTR
+         * because pipeline protocols assume that both peers use the same
+         * fragment sizes (and they are different for different memory types by
+         * default). */
+        frag_mem_type = init_params->rkey_config_key->mem_type;
```

Comment:
what happens if the memory type received in the RTR is not of a fragment, but belongs to a user buffer?

Reply:
the code would still be correct. If it is cuda memory, cuda fragment will be selected (which is perf efficient). Otherwise, host chunk will be used (also beneficial on receiver)